### PR TITLE
Feature: Add check for ledgers received by TmStatusChange.

### PR DESCRIPTION
**Pull request by lwojtow**

### CODE REVIEW

File: SPEC.md [hunk 1]
```diff
@@ -150,3 +150,9 @@ The test index makes use of symbolic language in describing connection and messa
     <- mtPROPOSE_LEDGER with node's node_pub_key
     -> mtSQUELCH (node_pub_key)
     <- mtPROPOSE_LEDGER with node's node_pub_key
+
+### ZG-CONFORMANCE-010
+
+    The node should send TmStatusChange message containing ledger info.
```
Comment (ljedrz) [inline SPEC.md:156]: on its own? the conditions are unclear

File: SPEC.md [hunk 2]
```diff
@@ -150,3 +150,9 @@ The test index makes use of symbolic language in describing connection and messa
...
+    <>
+    <- TmStatusChange with correct ledger hash and sequence
```
Comment (ljedrz) [inline SPEC.md:158]: what makes them correct?

File: src/tests/conformance/status.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        config::TestConfig,
+        constants::EXPECTED_RESULT_TIMEOUT,
+        rpc::{get_ledger_info, LedgerInfoResponse, RpcResponse},
+        synth_node::SyntheticNode,
+    },
+};
+
+#[tokio::test]
+#[allow(non_snake_case)]
+async fn C010_TM_STATUS_CHANGE_should_receive_ledger_information_by_status_change() {
```
Comment (ljedrz) [inline src/tests/conformance/status.rs:19]: this name is unclear, we need something referring to the node

File: src/tests/conformance/status.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    wait_for_status_change(&mut sn, rpc_ledger_index, rpc_ledger_hash)
+        .await
+        .expect("no status message within required time");
+
+    // Cleanup.
+    node.stop().expect("unable to stop stateful node");
+    sn.shut_down().await;
+}
+
+async fn wait_for_status_change(
```
Comment (ljedrz) [inline src/tests/conformance/status.rs:57]: don't we already have some helper function to wait for some specific message and return it?

File: src/tests/conformance/status.rs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                if ledger_seq == rpc_ledger_index && ledger_hash == rpc_ledger_hash {
+                    break;
+                }
+            }
+        }
+    })
+    .await
+}
+
+fn parse_ledger_index(info: &RpcResponse<LedgerInfoResponse>) -> Result<u32, ParseIntError> {
```
Comment (ljedrz) [inline src/tests/conformance/status.rs:80]: I don't think this helper is especially useful, it's called just once

File: src/tests/conformance/status.rs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    info.result.ledger.ledger_index.parse::<u32>()
+}
+
+async fn wait_for_ledger_info_by_rpc() -> Result<RpcResponse<LedgerInfoResponse>, Elapsed> {
```
Comment (ljedrz) [inline src/tests/conformance/status.rs:84]: same point as in https://github.com/runziggurat/xrpl/pull/48/files#r982373504

File: src/tools/rpc.rs [hunk 7]
```diff
@@ -73,6 +73,42 @@ pub async fn get_transaction_info(
...
+    let response = build_json_request(&build_ledger_info_request())
+        .send()
+        .await?;
+    Ok(response
+        .error_for_status()?
+        .json::<RpcResponse<LedgerInfoResponse>>()
+        .await?)
+}
+
+fn build_ledger_info_request() -> RpcRequest<Vec<LedgerInfoRequest>> {
```
Comment (ljedrz) [inline src/tools/rpc.rs:86]: this helper doesn't seem practical either, and the `builder` part in the name is misleading (I was expecting the builder pattern)

**Review by ljedrz — COMMENTED** [review]
Looks like a useful test case; I left some comments; also, we could use a more informative PR description, as the title isn't self-explanatory.

File: src/tests/conformance/status.rs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        .expect("unable to decode ledger hash");
+
+    // Wait for TmStatusChange message.
+    wait_for_status_change(&mut sn, rpc_ledger_index, rpc_ledger_hash)
+        .await
+        .expect("no status message within required time");
+
+    // Cleanup.
+    node.stop().expect("unable to stop stateful node");
+    sn.shut_down().await;
```
Comment (Rqnsom) [inline src/tests/conformance/status.rs:54]: We should reverse the order here.

```suggestion
    sn.shut_down().await;
    node.stop().expect("unable to stop stateful node");
```

Comment (Rqnsom) [inline src/tests/conformance/status.rs:54]: We should reverse the order here.

```suggestion
    sn.shut_down().await;
    node.stop().expect("unable to stop stateful node");
```

File: SPEC.md (hunk 1 above)
Comment (lwojtow) [reply SPEC.md:156]: Added more information to README.

File: SPEC.md (hunk 2 above)
Comment (lwojtow) [reply SPEC.md:158]: Added more information to README.

File: src/tests/conformance/status.rs (hunk 3 above)
Comment (lwojtow) [reply src/tests/conformance/status.rs:19]: Changed the name of the test.

File: src/tools/rpc.rs (hunk 7 above)
Comment (lwojtow) [reply src/tools/rpc.rs:86]: Removed

File: src/tests/conformance/status.rs (hunk 4 above)
Comment (lwojtow) [reply src/tests/conformance/status.rs:57]: Done

File: src/tests/conformance/status.rs (hunk 6 above)
Comment (lwojtow) [reply src/tests/conformance/status.rs:84]: Not for RPC. We could add this at some point, though.

File: SPEC.md [hunk 9]
```diff
@@ -150,3 +150,12 @@ The test index makes use of symbolic language in describing connection and messa
     <- mtPROPOSE_LEDGER with node's node_pub_key
     -> mtSQUELCH (node_pub_key)
     <- mtPROPOSE_LEDGER with node's node_pub_key
+
+### ZG-CONFORMANCE-010
+
+    The node should send TmStatusChange message containing ledger info.
+    This message should be sent without requesting explicitly.
+    To ensure ledger correctness, tests asks for ledger information via RPC 
+    and compare results with TmStatusChange payload.
```
Comment (ljedrz) [inline SPEC.md:159]:
```suggestion
    and compares the results with the TmStatusChange payload. 
```

File: SPEC.md [hunk 10]
```diff
@@ -150,3 +150,12 @@ The test index makes use of symbolic language in describing connection and messa
     <- mtPROPOSE_LEDGER with node's node_pub_key
     -> mtSQUELCH (node_pub_key)
     <- mtPROPOSE_LEDGER with node's node_pub_key
+
+### ZG-CONFORMANCE-010
+
+    The node should send TmStatusChange message containing ledger info.
+    This message should be sent without requesting explicitly.
```
Comment (ljedrz) [inline SPEC.md:157]:
```suggestion
    This message should be sent without any explicit requests.
```

File: SPEC.md [hunk 11]
```diff
@@ -150,3 +150,12 @@ The test index makes use of symbolic language in describing connection and messa
...
+    To ensure ledger correctness, tests asks for ledger information via RPC
```
Comment (ljedrz) [inline SPEC.md:158]:
```suggestion
    To ensure ledger correctness, the test asks for ledger information via RPC first 
```

File: SPEC.md [hunk 12]
```diff
@@ -150,3 +150,12 @@ The test index makes use of symbolic language in describing connection and messa
...
     <- mtPROPOSE_LEDGER with node's node_pub_key
+
+### ZG-CONFORMANCE-010
+
+    The node should send TmStatusChange message containing ledger info.
+    This message should be sent without requesting explicitly.
+    To ensure ledger correctness, tests asks for ledger information via RPC 
+    and compare results with TmStatusChange payload. 
+    <>
+    <- TmStatusChange with correct ledger hash and sequence
```
Comment (ljedrz) [inline SPEC.md:161]:
```suggestion
    <- TmStatusChange with the correct ledger hash and sequence
```

File: src/tests/conformance/status.rs [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+    // Create a stateful node.
+    let mut node = build_stateful_builder(target.path().to_path_buf())
+        .expect("unable to get stateful builder")
+        .start(false)
+        .await
+        .expect("unable to start stateful node");
+
+    // Connect synth node.
+    let mut sn = SyntheticNode::new(&TestConfig::default()).await;
```
Comment (ljedrz) [inline src/tests/conformance/status.rs:33]:
```suggestion
    let mut sn = SyntheticNode::new(&Default::default()).await;
```

File: src/tests/conformance/status.rs [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    sn.connect(node.addr()).await.unwrap();
+
+    // Get ledger information via RPC.
+    let info = wait_for_ledger_info_by_rpc()
+        .await
+        .expect("no ledger info within required time");
```
Comment (ljedrz) [inline src/tests/conformance/status.rs:39]:
```suggestion
        .expect("no ledger info within the specified time limit");
```

File: src/tests/conformance/status.rs [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    // Wait for TmStatusChange message.
+    let check = |m: &BinaryMessage| {
+        matches!(
+            &m.payload,
+            Payload::TmStatusChange(TmStatusChange {
+                ledger_seq: Some(ledger_seq),
+                ledger_hash: Some(ledger_hash),
+                ..
+            })
+            if ledger_seq == &rpc_ledger_index && ledger_hash.as_slice() == rpc_ledger_hash
```
Comment (ljedrz) [inline src/tests/conformance/status.rs:64]:
```suggestion
            ledger_seq == &rpc_ledger_index && ledger_hash.as_slice() == rpc_ledger_hash
```
do double-check that this test case compiles, this doesn't seem right

File: src/tests/conformance/status.rs [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    node.stop().expect("unable to stop stateful node");
+}
+
+async fn wait_for_ledger_info_by_rpc() -> Result<RpcResponse<LedgerInfoResponse>, Elapsed> {
+    timeout(EXPECTED_RESULT_TIMEOUT, async {
+        loop {
+            if let Ok(info) = get_ledger_info().await {
+                break info;
+            } else {
+                sleep(Duration::from_secs(1)).await;
```
Comment (ljedrz) [inline src/tests/conformance/status.rs:80]: why so long? I propose cutting it to 200-250ms

File: SPEC.md (hunk 9 above)
Comment (lwojtow) [reply SPEC.md:159]: Done

File: src/tests/conformance/status.rs (hunk 16 above)
Comment (lwojtow) [reply src/tests/conformance/status.rs:80]: Will do. But it takes about 6 secs for the node to get ready, so we won't really save any time here.

File: src/tests/conformance/status.rs (hunk 14 above)
Comment (lwojtow) [reply src/tests/conformance/status.rs:39]: Done

File: src/tests/conformance/status.rs (hunk 13 above)
Comment (lwojtow) [reply src/tests/conformance/status.rs:33]: Done

File: SPEC.md (hunk 12 above)
Comment (lwojtow) [reply SPEC.md:161]: Done

File: src/tests/conformance/status.rs (hunk 6 above)
Comment (lwojtow) [reply src/tests/conformance/status.rs:74]: Moved to RPC module to keep consistent with other similar methods.

File: src/tests/conformance/squelch.rs [hunk 17]
```diff
@@ -46,7 +46,7 @@ async fn c009_TM_SQUELCH_cannot_squelch_peer_ledger_proposals() {
     let target = TempDir::new().expect("unable to create TempDir");
     let mut node = build_stateful_builder(target.path().to_path_buf())
         .expect("unable to get stateful builder")
-        .start(true)
+        .start(false)
```
[the remark this answers is not in the document]
Comment (lwojtow) [reply src/tests/conformance/squelch.rs:49]: Done

**Review by ljedrz — APPROVED** [review]
LGTM; @niklaslong?

**Review by niklaslong — APPROVED** [review]

**Pull request merged**